contracts/common/GetActionSummary.sol

```diff
@@ -0,0 +1,203 @@
+}
+
+contract GetActionSummary is ExtractCallData, GetActionDomainSkillId {
+  using Bytes4Includes for bytes4[];
+
+  bytes4 constant MULTICALL = bytes4(keccak256("multicall(bytes[])"));
+  bytes4 constant SET_EXPENDITURE_STATE =
+    bytes4(
+      keccak256("setExpenditureState(uint256,uint256,uint256,uint256,bool[],bytes32[],bytes32)")
+    );
+  bytes4 constant SET_EXPENDITURE_PAYOUT =
+    bytes4(keccak256("setExpenditurePayout(uint256,uint256,uint256,uint256,address,uint256)"));
+
+  bytes32 constant ROOT_ROLES = ((bytes32(uint256(1)) <<
+    uint8(ColonyDataTypes.ColonyRole.Recovery)) |
+    (bytes32(uint256(1)) << uint8(ColonyDataTypes.ColonyRole.Root)));
+
+  bytes32 constant ONLY_ROOT_ROLE_MASK =
+    bytes32(uint256(1)) << uint8(ColonyDataTypes.ColonyRole.Root);
+
+  bytes4 constant SET_USER_ROLES =
+    bytes4(keccak256("setUserRoles(uint256,uint256,address,uint256,bytes32)"));
+
+  bytes32 constant ONLY_ARCHITECTURE_ROLE_MASK =
+    bytes32(uint256(1)) << uint8(ColonyDataTypes.ColonyRole.Architecture);
+
+  function getExpenditureId(bytes memory action) internal pure returns (uint256 expenditureId) {
+    bytes4 sig = getSig(action);
+    assert(isExpenditureSig(sig));
+
+    assembly {
+      expenditureId := mload(add(action, 0x64))
+    }
+  }
+
+  function getSingleActionSummary(
+    address colonyNetworkAddress,
+    address colonyAddress,
+    bytes memory _action,
+    address _altTarget
+  ) public view returns (ActionSummary memory) {
+    bytes4 sig = getSig(_action);
+    require(sig != MULTICALL, "colony-get-action-summary-no-nested-multicalls");
+
+    // TODO: explicitly check `isExtension` for target, currently this simply errors
+    address target = getTarget(_altTarget, colonyAddress);
+    bytes32 permissionMask = ColonyRoles(target).getCapabilityRoles(sig);
+
+    ActionSummary memory summary;
+    summary.sig = sig;
+
+    if (isExpenditureSig(sig)) {
+      summary.domainSkillId = getActionDomainSkillId(_action, colonyNetworkAddress, colonyAddress);
+      summary.expenditureId = getExpenditureId(_action);
+    } else if (permissionMask | ROOT_ROLES == ROOT_ROLES) {
+      summary.domainSkillId = IColony(colonyAddress).getDomain(1).skillId;
+      if (permissionMask == bytes32(0)) {
+        summary.requiredPermissions = ONLY_ROOT_ROLE_MASK;
+      } else {
+        summary.requiredPermissions = permissionMask;
+      }
+    } else {
+      summary.domainSkillId = getActionDomainSkillId(_action, colonyNetworkAddress, colonyAddress);
+      // A special case for permission-setting functions, which can be called by root (in root) and
+      // by architecture (if being used in a child domain of where you have the permission)
+      if (permissionMask == (ONLY_ARCHITECTURE_ROLE_MASK | ONLY_ROOT_ROLE_MASK)) {
+        // slither-disable-next-line incorrect-equality
+        if (summary.domainSkillId == IColony(colonyAddress).getDomain(1).skillId) {
+          permissionMask = ONLY_ROOT_ROLE_MASK;
```

Comment:
Ah, I see that the issue I raised elsewhere is being handled here -- for the role-setting functions, only the root role is allowed when calling the function in the root domain.